Add Products of Inertia to RigidBody3D

erikscott128 opened the pull request:
## Current Situation
Currently users can only define custom inertia parallel to the body axes of a `RigidBody3D` node. That is, the inertia's principle axes are assumed to be aligned with the body axes, and only the principle moments of inertia (Ixx, Iyy, Izz) can be defined. 

## Improvements Made
This branch adds the ability to set the off diagonal products of inertia (Iyz, Ixz, Ixy) such that the entire local inertia tensor can be defined without the ability for the user to accidentally define an invalid, non-symmetrical matrix. The functionality is implemented for both the GodotPhysics3D and Jolt Physics engines. Behavior in each engine has been shown to be identical. The products of inertia are ordered (Iyz, Ixz, Ixy). If combined with the on-axis `inertia` value, the 6-element array is consistent with Voigt notation for representing symmetric tensors. 

## Benefits
Bodies with non-zero products of inertia will respond differently to external moments. For instance, applying a pure x-axis moment to a body with non-zero Ixz and/or Ixy values will result in rotation not purely about the x-axis. This allows for realistic behavior of physics objects, such as vehicles, with one or more off-axis inertia terms.  

## Implementation Details
The products of inertia are passed to `GodotBody3D` or `JoltBody3D` in the same way as the on-axis `inertia`. In the Jolt engine, a full local inertia tensor is already provided as part of the `Mass Properties` object, and the inertia values need only be set to the correct elements. In the Godot physics engine, this functionality had to be added. A local inertia tensor object now exists within the `GodotBody3D` class and the inertia values are assigned to the proper elements. This matrix is then decomposed into its eigenvalues (principle moments of inertia) and eigenvectors (principle axes of inertia) for use within the existing Godot implementation. The algorithm for this was copied from the Jolt physics engine library but modified to use Godot's `Vector3` and `Basis` classes. The algorithm is an implementation of the [Jacobi eigenvalue algorithm](https://en.wikipedia.org/wiki/Jacobi_eigenvalue_algorithm), which only works for real-symmetric matrices. 

The appropriate doc .xml files have been updated to reflect the new `product_of_inertia` field, and the addition of `BODY_PARAM_PROD_INERTIA` to the `BodyParam` enum.  

## Anticipated Discussion Points
- I chose to implement the "alternate" convention, where the products of inertia are negated before being inserted into the tensor. This may be controversial. See https://en.wikipedia.org/wiki/Moment_of_inertia#Inertia_tensor for more info.
- The `eigen_value_symmetric` function is currently implemented in a new, separate file. This should probably be moved to either the `Basis` class as a method, or to the math_funcs.h/.cpp files. I defer this decision to someone more familiar with how Godot is and should be laid out. It duplicates the `diagonalize` method already present in the `Basis` class, except that it also returns the eigenvalues, which are necessary.

erikscott128 committed:
Product of inertia UI inputs

Added UI inputs for product of inertia

erikscott128 committed:
Initial Jolt implementation

Initial jolt implementation of products of inertia

erikscott128 committed:
Products of Inertia implemented for both physics engines

Products of inertia have been added and implemented for both GodotPhysics3D and JoltPhysics engines.

erikscott128 committed:
Update RigidBody3D.xml to include product_of_inertia

Added documentation for the Product of Inertia parameter in RigidBody3D

erikscott128 committed:
Merge branch 'godotengine:master' into products_of_inertia

erikscott128 committed:
Misc cleanup

Remove unused `use_products_of_inertia` flag. Added enum to doc.

erikscott128 committed:
Merge branch 'products_of_inertia' of https://github.com/erikscott128/godot into products_of_inertia

erikscott128 committed:
Updated file header

Updated the file header for eigen_value_symmetric.h to the proper Godot format.

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,144 @@
+
+#pragma once
+
+#include "core/math/vector3.h"
```
```suggestion
```
Redundant

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,144 @@
+        real_t sm = 0.0;
+        for (int ip = 0; ip < N; ++ip){
+            for (int iq = ip +1; iq < N; ++iq){
+                sm += abs(a[ip][iq]);
```
```suggestion
                sm += Math::abs(a[ip][iq]);
```
Please use the built in methods, for all

jrouwe commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,144 @@
+					eigval_p -= h;
+					eigval_q += h;
+
+                    #define JPH_EVS_ROTATE(a, i, j, k, l)		\
```
Some Jolt naming convention leftovers: `JPH_...`

jrouwe commented on modules/godot_physics_3d/godot_body_3d.cpp:
```diff
@@ -320,7 +366,8 @@ void GodotBody3D::set_mode(PhysicsServer3D::BodyMode p_mode) {
 			if (!calculate_inertia) {
 				principal_inertia_axes_local = Basis();
-				_inv_inertia = inertia.inverse();
+				// _inv_inertia = inertia.inverse();
```
Leftover comment

erikscott128 replied:
Sorry about that

erikscott128 committed:
Code style cleanup

Cleanup of code using clang-format

erikscott128 committed:
Added protection to eigenval calculation

Added protection to the eigenvalue/vector calculation such that, if it fails, the principal axes revert to the body-aligned frame, and the on-axis moments of inertia are used.

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+		real_t sm = 0.0;
+		for (int ip = 0; ip < N; ++ip) {
+			for (int iq = ip + 1; iq < N; ++iq) {
+				sm += abs(a[ip][iq]);
```
```suggestion
				sm += Math::abs(a[ip][iq]);
```
For all, please use the built-in methods

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include "core/math/basis.h"
+#include "core/math/vector3.h"
```
```suggestion
```
Already included

erikscott128 committed:
Eigen calc cleanup

Added warning if unable to find principle axes. Updated #define name to GODOT_EVS_ROTATE. Included math_funcs library and added Math namespace to abs functions.

erikscott128 committed:
Bind enum and fix doc

Added `BIND_ENUM_CONSTANT(BODY_PARAM_PROD_INERTIA);` to `PhysicsServer3D`. Updated RigidBody3D.xml to put product_of_inertia in proper alphabetical order.

erikscott128 committed:
Corrected enum order

New enum must be added to end of list (but before max). Doc updated.

aaronfranke commented on servers/physics_3d/physics_server_3d.cpp:
```diff
@@ -1091,6 +1091,7 @@ void PhysicsServer3D::_bind_methods() {
 	BIND_ENUM_CONSTANT(BODY_PARAM_FRICTION);
 	BIND_ENUM_CONSTANT(BODY_PARAM_MASS);
 	BIND_ENUM_CONSTANT(BODY_PARAM_INERTIA);
+	BIND_ENUM_CONSTANT(BODY_PARAM_PROD_INERTIA);
```
This should also be moved down to keep the binding code in the same order.

erikscott128 replied:
Just made the edit on my end. It'll get pushed with the next commit.

erikscott128 committed:
Corrected enum bind order

Changed the order in which the enum entries are bound to reflect new order.

erikscott128 commented:
> Until you created this PR I had never seen anyone use the 'products of inertia' notation. I've always stored inertia as a rotation + diagonal or just plain as a tensor. 

Separating the inertia entries into two separate 3-element vectors wouldn't have been my ideal implementation if I were designing this from the ground-up. It's the best implementation I could conceive of while keeping the existing structure in place to maintain compatibility with existing projects. I didn't want to expose the raw tensor because I didn't want uninitiated users to treat it like a basis and enter a non-symmetrical matrix. This implementation means it can be safely ignored by the majority of users who don't care, while providing a decent interface for those who do. The choice of using Voigt notation was driven simply by having to choose how to order these elements and I turned to an already-established convention. To my knowledge, Voigt notation is never actually used in dynamics to represent inertia; it's something structural engineers would use to represent stress and strain tensors, which are similarly symmetric, but benefit from the 1-dimensional representation in actual calculation. 

> Specifying the components of the inertia tensor directly is probably not understandable for most people, so if we use the 'normal' or 'alternate' version doesn't seem like a big deal to me. If we want a more user friendly version then I would suggest allowing the user to specify a rotation instead + a way of visualizing the resulting inertia tensor (but I can imagine that would complicate things in your sim).

While I think you're right that there is some intuition behind exposing it as principle inertia and a principle basis (or rotation), in practice this actually complicates things. The way the inertia is actually computed in most cases is in a coordinate system that's already known (ie, the body-attached, or "local' frame). The Inertia tensor is built up from a variety of known masses within that coordinate system plus their individual inertias (if not modeled as point masses). As an example, here's how I model a simple fuel burn in GDScript using the additions from this pull request: 

```
extends RigidBody3D

var fuel = 20;  # kg
var fuel_loc : Vector3 = Vector3(0.5,0.5,0.0)
var mass_empty : float = 20.0;
var cg_empty : Vector3 = Vector3(0.0, 0.0, 0.0)
var inertia_empty : Vector3 = mass_empty*Vector3(1.0,1.0,1.0)/12
var product_of_inertia_empty : Vector3 = Vector3(0.0, 0.0, 0.0)

var fuel_burn_rate : float = 1  # kg/s

func _physics_process(delta: float) -> void:
	
	var mass_tot : float = fuel+mass_empty
	var cg : Vector3 = (mass_empty*cg_empty + fuel*fuel_loc)/(mass_tot)
	
	var fuel_loc_from_cg : Vector3 = fuel_loc - cg
	
	# Calculate inertia for current fuel (assume fuel is a point mass)
	var inertia_tot : Vector3 = inertia_empty
	inertia_tot.x += fuel * (fuel_loc_from_cg.y**2 + fuel_loc_from_cg.z**2)
	inertia_tot.y += fuel * (fuel_loc_from_cg.x**2 + fuel_loc_from_cg.z**2)
	inertia_tot.z += fuel * (fuel_loc_from_cg.x**2 + fuel_loc_from_cg.y**2)
	
	var product_of_inertia_tot : Vector3 = product_of_inertia_empty
	product_of_inertia_tot.x += fuel * (fuel_loc_from_cg.y * fuel_loc_from_cg.z)
	product_of_inertia_tot.y += fuel * (fuel_loc_from_cg.x * fuel_loc_from_cg.z)
	product_of_inertia_tot.z += fuel * (fuel_loc_from_cg.x * fuel_loc_from_cg.y)
	
	# set mass properties (this is where a set_mass_properties function would be useful)
	set_mass(mass_tot)
	set_center_of_mass(cg)
	set_inertia(inertia_tot)
	set_product_of_inertia(product_of_inertia_tot)
	
	# burn fuel
	fuel -= fuel_burn_rate*delta;
	fuel = maxf(fuel, 0.0)
```

As you can see, this is actually quite easy to do. It also supports my choice of Voigt order because there's some nice consistency in the expressions (The x term in both vectors is dependent on the y and z coordinates of the mass, y on x and z, and z on x and y). And not having to remember to stick a negative sign in there is helpful as well, which supports the chosen sign convention, so I think I'm sticking with it (just got to make sure that's well documented). 

> Added a reply there.

While this is probably not the right place to discuss this, I think having a project setting to enable full calculation of the rotational dynamics is a good idea. I replied in the issue as well, and it can be discussed in the next pull request. 

I appreciate the feedback!

erikscott128 committed:
Initial working set_mass_properties commit

Initial working implementation of the `set_mass_properties` function which allows the user to set all mass properties at once without redundant calculations.

erikscott128 committed:
Doc updated

Updated documentation to include new `set_mass_properties` function

erikscott128 committed:
Merge pull request #1 from erikscott128/consolidated_mass_prop_func

Consolidated mass properties function

AThousandShips commented on servers/physics_3d/physics_server_3d_wrap_mt.h:
```diff
@@ -220,6 +220,7 @@ class PhysicsServer3DWrapMT : public PhysicsServer3D {
 	FUNC3(body_set_param, RID, BodyParameter, const Variant &);
 	FUNC2RC(Variant, body_get_param, RID, BodyParameter);
 
+	FUNC5(body_set_mass_properties, RID, const real_t &, const Vector3 &, const Vector3 &, const Vector3 &);
```
```suggestion
	FUNC5(body_set_mass_properties, RID, real_t, const Vector3 &, const Vector3 &, const Vector3 &);
```

erikscott128 committed:
Small correction to doc

AThousandShips commented on doc/classes/PhysicsServer3D.xml:
```diff
@@ -593,6 +593,16 @@
+			<param index="2" name="center_of_mass" type="Vector3" />
+			<param index="3" name="inertia" type="Vector3" />
+			<param index="4" name="product_of_inertia" type="Vector3" />
+			<description>
```
Needs a description

erikscott128 committed:
Corrected doc again

Added doc entry for `body_set_mass_properties` to PhysicsServer3DExtension.xml, not PhysicsServer3D.xml.

erikscott128 committed:
Exposed Jolt's Gyroscopic Forces Option

Added entry to Jolt's project settings that enables setting the `mApplyGyroscopicForce` option.

erikscott128 committed:
GodotPhysics3D initial implementation of gyro forces

Initial implementation of "gyroscopic forces" in the Godot 3D physics engine.

erikscott128 committed:
Fix method bind

Accidentally bound incorrect method (corrected `body_reset_mass_properties` to `body_set_mass_properties`)

erikscott128 committed:
Doc updated

Updated ProjectSettings and PhysicsServer3D docs

erikscott128 committed:
Update PhysicsServer3D.xml

jrouwe commented on modules/godot_physics_3d/godot_body_3d.cpp:
```diff
@@ -320,7 +399,7 @@ void GodotBody3D::set_mode(PhysicsServer3D::BodyMode p_mode) {
 			_inv_mass = mass > 0 ? (1.0 / mass) : 0;
 			if (!calculate_inertia) {
```
Should this trigger a warning when you set the product of inertia without having set the inertia first?

erikscott128 replied:
That's a good idea. Noted.

jrouwe commented on modules/jolt_physics/objects/jolt_body_3d.cpp:
```diff
@@ -690,6 +707,28 @@ void JoltBody3D::set_param(PhysicsServer3D::BodyParameter p_param, const Variant
 }
 
+void JoltBody3D::set_mass_properties(real_t p_mass, const Vector3 &p_center_of_mass, const Vector3 &p_inertia, const Vector3 &p_product_of_inertia) {
+	if (p_mass != mass) {
```
The ifs that only assign a value in this function are redunant (3x), just assigning a value is cheaper in most cases.

erikscott128 commented on modules/godot_physics_3d/godot_body_3d.cpp:
```diff
@@ -53,6 +83,34 @@ void GodotBody3D::_update_transform_dependent() {
 
+void GodotBody3D::set_mass_properties(real_t p_mass, const Vector3 &p_center_of_mass, const Vector3 &p_inertia, const Vector3 &p_product_of_inertia) {
+	// Set Mass
+	real_t mass_value = p_mass;
```
Yes, not sure how that happened. I'll fix it.

erikscott128 committed:
Minor code-review changes

jrouwe reviewed (approved):
Don't have any authority here, but I think it's good like this!

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,146 @@
+						t = a_pq / h;
+					} else {
+						real_t theta = 0.5 * h / a_pq; // Warning: Can become infinite if a(ip, iq) is very small which may trigger an invalid float exception
+						t = 1.0 / (Math::abs(theta) + sqrt(1.0 + theta * theta)); // If theta becomes inf, t will be 0 so the infinite is not a problem for the algorithm
```
```suggestion
						t = 1.0 / (Math::abs(theta) + Math::sqrt(1.0 + theta * theta)); // If theta becomes inf, t will be 0 so the infinite is not a problem for the algorithm
```

erikscott128 replied:
Fixed in latest commit.

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,144 @@
+
+// This code is simply a re-implementation of the same function from Jolt Physics, 
+// but using Godot's Vector3 and Basis structures
+bool eigen_value_symmetric(const Basis &in_matrix, Basis &out_eig_vec, Vector3 &out_eig_val) {
```
```suggestion
bool eigen_value_symmetric(const Basis &p_in_matrix, Basis &r_out_eig_vec, Vector3 &r_out_eig_val) {
```

erikscott128 replied:
I will name these `p_matrix`, `r_eig_vec` and `r_eig_val`. The `in_` and `out_` are redundant with the new prefixes. 
Subsequent line `Basis a = in_matrix;` will become `Basis a = p_matrix;`.

AThousandShips commented on modules/godot_physics_3d/godot_body_3d.cpp:
```diff
@@ -41,6 +42,34 @@ void GodotBody3D::_mass_properties_changed() {
+	inertia_tensor_local[1][0] = -product_of_inertia.z;
+
+	principal_inertia_axes_local = Basis();
+	Vector3 principal_inertia = Vector3();
```
```suggestion
	Vector3 principal_inertia;
```

erikscott128 replied:
Fixed in latest commit.

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+
+	Basis a = in_matrix;
+
+	Vector3 b, z;
```
```suggestion
	Vector3 b;
	Vector3 z;
```
Please keep one declaration per line

erikscott128 replied:
Fixed in latest commit.

AThousandShips commented on servers/physics_3d/physics_server_3d.h:
```diff
@@ -449,12 +449,14 @@ class PhysicsServer3D : public Object {
 	virtual void body_set_param(RID p_body, BodyParameter p_param, const Variant &p_value) = 0;
 	virtual Variant body_get_param(RID p_body, BodyParameter p_param) const = 0;
 
+	virtual void body_set_mass_properties(RID p_body, const real_t &p_mass, const Vector3 &p_center_of_mass, const Vector3 &p_inertia, const Vector3 &p_product_of_inertia) = 0;
```
```suggestion
	virtual void body_set_mass_properties(RID p_body, real_t p_mass, const Vector3 &p_center_of_mass, const Vector3 &p_inertia, const Vector3 &p_product_of_inertia) = 0;
```

erikscott128 replied:
Updated to pass by value in GodotPhysicsServer3D and JoltPhysicsServer3D as well.

AThousandShips commented on servers/physics_3d/physics_server_3d_wrap_mt.h:
```diff
@@ -220,6 +220,7 @@ class PhysicsServer3DWrapMT : public PhysicsServer3D {
 	FUNC3(body_set_param, RID, BodyParameter, const Variant &);
 	FUNC2RC(Variant, body_get_param, RID, BodyParameter);
 
+	FUNC5(body_set_mass_properties, RID, const real_t &, const Vector3 &, const Vector3 &, const Vector3 &);
```
```suggestion
	FUNC5(body_set_mass_properties, RID, real_t, const Vector3 &, const Vector3 &, const Vector3 &);
```

erikscott128 replied:
Fixed in latest commit. Also applied to GodotPhysicsServer3D and JoltPhysicsServer3D.

erikscott128 committed:
Code review changes (AThousandShips)

Code review changes in response to AThousandShips (thank you!): eigen_value_summetric cleaned up to adhere to godot style guidelines for input and return parameters. Mass is now passed by value in set_mass_properties. Other minor issues corrected.

erikscott128 commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include "core/math/basis.h"
+#include "core/math/vector3.h"
```
Fixed in latest commit. For my own clarification, is it considered good practice to not include something if another header file already includes it? I would have thought it would be good to explicitly include things you need and let header guards handle any duplication.

erikscott128 commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include "core/math/basis.h"
+#include "core/math/math_funcs.h"
```
Fixed in latest commit.

erikscott128 commented on servers/physics_3d/physics_server_3d_extension.h:
```diff
@@ -332,6 +332,7 @@ class PhysicsServer3DExtension : public PhysicsServer3D {
 	EXBIND3(body_set_param, RID, BodyParameter, const Variant &)
 	EXBIND2RC(Variant, body_get_param, RID, BodyParameter)
 
+	EXBIND5(body_set_mass_properties, RID, const real_t &, const Vector3 &, const Vector3 &, const Vector3 &)
```
Fixed in latest commit. Also applied to GodotPhysicsServer3D and JoltPhysicsServer3D.

erikscott128 committed:
Merge branch 'godotengine:master' into correct_rotational_motion

erikscott128 committed:
Merge branch 'godotengine:master' into products_of_inertia

erikscott128 committed:
Merge branch 'correct_rotational_motion' into products_of_inertia

erikscott128 committed:
Revert "Merge branch 'correct_rotational_motion' into products_of_inertia"

This reverts commit 814cbbaeb8d0319e9107919209a59fa8fb400617, reversing
changes made to ab0095cc61625eec8f1b6cb0f50a4e715e6f8f5d.

AThousandShips commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+							t = -t;
+						}
+					}
+					real_t c = 1.0 / sqrt(1.0 + t * t);
```
```suggestion
					real_t c = 1.0 / Math::sqrt(1.0 + t * t);
```

erikscott128 committed:
Update core/math/eigen_value_symmetric.h

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

aaronfranke commented on core/math/eigen_value_symmetric.h:
```diff
@@ -0,0 +1,145 @@
+	Vector3 b;
+	Vector3 z;
+
+	for (int ip = 0; ip < N; ++ip) {
```
We generally use postfix `++` in Godot.

Cykyrios commented on doc/classes/RigidBody3D.xml:
```diff
@@ -233,6 +243,11 @@
+			The body's products of inertia in (Iyz, Ixz, Ixy) order. Like the inertia parameter, this also determines how much torque it takes to rotate an object. However, it describes how asymmetrically the mass is distributed relative to the body's axes. When combined with the inertia vector, the local inertia tensor can be fully defined. However, for most applications, this can safely be left at zero.
+			If inertia is computed automatically, the value set here will be ignored.
+			[b]Note:[/b] The "alternate" convention for products of inertia is used. These values will be negated in the inertia tensor.
+		</member>
```
You can use either `[url]https://example[/url]` or `[url=https://example]text to display[/url]` for this. If you ever need to link to another page of the Godot documentation, you should use `[url=$DOCS_URL/page]`, so the proper version can be determined.

erikscott128 replied:
Documentation edits have been made to be more descriptive and include a link to the Wikipedia article. I have included example code that demonstrates how to compute the local inertia tensor from the principle inertia. I also updated the existing inertia description to include the Wikipedia link and to explain its role in the overall inertia tensor.

erikscott128 committed:
Update eigen_value_symmetric.h

Corrected for-loop increments to be postfix

erikscott128 committed:
Update RigidBody3D.xml

Updated documentation for both `inertia` and `product_of_inertia` to be more specific and better describe how they contribute to the overall inertia tensor. Also included example code for how to translate from the principle inertia representation.

aaronfranke commented on doc/classes/RigidBody3D.xml:
```diff
@@ -244,9 +245,28 @@
+			func InertiaFromPrincipal(Vector3 principalInertia, Basis principalAxes)
+			{
+				Basis principalInertiaTensor = Basis.FromScale(principalInertia);
+				Basis inertiaTensor  = principalAxes * principalInertiaTensor * principalAxes.Transposed();
```
```suggestion
				var principal_inertia_tensor: Basis = Basis.from_scale(principal_inertia)
				var inertia_tensor: Basis = principal_axes * principal_inertia_tensor * principal_axes.transposed()
				inertia = Vector3(inertia_tensor.x.x, inertia_tensor.y.y, inertia_tensor.z.z)
				product_of_inertia = Vector3(-inertia_tensor.y.z, -inertia_tensor.x.z, -inertia_tensor.x.y)
			[/gdscript]
			[csharp]
			void InertiaFromPrincipal(Vector3 principalInertia, Basis principalAxes)
			{
				Basis principalInertiaTensor = Basis.FromScale(principalInertia);
				Basis inertiaTensor = principalAxes * principalInertiaTensor * principalAxes.Transposed();
```

erikscott128 committed:
Update doc/classes/RigidBody3D.xml

Minor example code formatting changes

Co-authored-by: Aaron Franke <arnfranke@yahoo.com>

erikscott128 committed:
Update RigidBody3D.xml

Minor code example formatting changes

erikscott128 committed:
Update RigidBody3D.xml

More corrections to example code
